src/hackerspace_online/adapter.py

```diff
@@ -1,8 +1,27 @@
 from allauth.account.adapter import DefaultAccountAdapter
 
+from allauth.socialaccount.adapter import DefaultSocialAccountAdapter
+from django.contrib.auth import get_user_model
+
+User = get_user_model()
+
 
 class CustomAccountAdapter(DefaultAccountAdapter):
 
     def clean_username(self, username, shallow=False):
         username = super().clean_username(username, shallow)
         return username.lower()
+
+
+class CustomSocialAccountAdapter(DefaultSocialAccountAdapter):
+
+    def pre_social_login(self, request, sociallogin):
```

Comment:
> Thank you for this! It looks great. Because it involves authentication and emails, I want to ensure we have all possibilities covered. I think we need a flow chart to cover all possible sign-up methods and email changing.
> 
> Here are a few things I tested that will be problems we need to handle.
> 
> * After logging in with Google, I could change my email address in my profile to anything.  The Google log-in still worked, but which email are notifications going to?
> * After logging in, I could change my email address in my profile to another user's Gmail address! (since they changed it to something else)
> * Does "remember me" still work with Google sign in? This is important.
> 
> Can we merge to two email fields? Having two seems like it will be a problem. Or maybe we just need tighter controls on changing the email address within a profile. I think ALL email addresses should be verified, can django-allauth handle that?




> After logging in with Google, I could change my email address in my profile to anything. The Google log-in still worked, but which email are notifications going to?

This will use the email in the profile. The google account is still associated to that account

> After logging in, I could change my email address in my profile to another user's Gmail address! (since they changed it to something else)

Yeah.. this isn't something we could control. Maybe we could disable changing emails? But we need to think about a workflow for this


> Can we merge to two email fields? Having two seems like it will be a problem. Or maybe we just need tighter controls on changing the email address within a profile. I think ALL email addresses should be verified, can django-allauth handle that?

What do you mean by merging? Also can you give me more details on ALL email addresses should be verified?

Reply:
> > After logging in, I could change my email address in my profile to another user's Gmail address! (since they changed it to something else)
> 
> Yeah.. this isn't something we could control. Maybe we could disable changing emails? But we need to think about a workflow for this

How does django-allauth handle this?  Currently the email field is the one on django's User model.  But I know allauth does email verification, so how do we use that? 

What are these settings actually doing in our project?

> Set ACCOUNT_EMAIL_VERIFICATION = "optional" so an email is sent for verifying emails after sign up
> Set ACCOUNT_UNIQUE_EMAIL = True to prevent duplicate emails since django-allauth also has their own way of handling emails


"an email is sent for verifying emails after sign up" <-- I haven't experienced this in any of the sign-up workflows (local or Google), so how do we use it?

Reply:
> > > After logging in, I could change my email address in my profile to another user's Gmail address! (since they changed it to something else)
> > 
> > 
> > Yeah.. this isn't something we could control. Maybe we could disable changing emails? But we need to think about a workflow for this
> 
> How does django-allauth handle this? Currently the email field is the one on django's User model. But I know allauth does email verification, so how do we use that?

django-allauth has a separate `EmailAddress` model that stores the one you used during registration. I think it would only start doing this after this setting `ACCOUNT_EMAIL_VERIFICATION = "optional"` was enabled


> What are these settings actually doing in our project?
> 
> > Set ACCOUNT_EMAIL_VERIFICATION = "optional" so an email is sent for verifying emails after sign up
> > Set ACCOUNT_UNIQUE_EMAIL = True to prevent duplicate emails since django-allauth also has their own way of handling emails
> 
> "an email is sent for verifying emails after sign up" <-- I haven't experienced this in any of the sign-up workflows (local or Google), so how do we use it?

From the docs:

> When set to “mandatory” the user is blocked from logging in until the email address is verified. Choose “optional” or “none” to allow logins with an unverified e-mail address. In case of “optional”, the e-mail verification mail is still sent, whereas in case of “none” no e-mail verification mails are sent.

Have you tried signing up as a new student? As you can see in this recording, a new tab opened after I signed up as a student.. I'm currently using [django-naomi](https://pypi.org/project/django-naomi/) to automatically open emails on browser but I think you could also see it inside the `_sent_mail` folder

https://user-images.githubusercontent.com/10972027/229258193-b4949e12-c39d-4c74-a38e-2d23cab18b6f.mp4

Reply:
@tylerecouture 

I created a somewhat guide (for myself as well) using [gherkin](https://cucumber.io/docs/gherkin/reference/) syntax regarding the sign in. Hope this makes sense and let me know what you think:

```cucumber
Feature: New User sign in with Google

  Scenario: A new student decides to sign in to a deck account using their Google Account
    Given a new user with a google account: student@gmail.com

    When a user clicks the google sign in button

    Then they should be redirected to the google sign in page and enter their credentials

    When they enter the correct credentials

    Then they should be redirected to the sign up page and they should enter the required fields and correct access code

    When they hit submit

    Then they should now be logged in
----

Feature: User Email Confirmation

  Background:
    A student previously signed in via google with account student@gmail.com

  Scenario: User changes their email from student@gmail.com to new_student@gmail.com

    Given a user that signed in via google with email: student@gmail.com

    When they go to the profile page and change their email to new_student@gmail.com

    Then they hit submit

    Then they should receive a confirmation email to their new email new_student@gmail.com

    Then their social account (Google Sign in) should still work even after they changed their email since a user who signed in using a social provider does not necessarily mean that they cannot use any more. The only way where they won't be able to sign in is when they disconnect their social account from their bytedeck account

    Then they should still be able to sign in via google using student@gmail.com google account
```


I've tested social login with stackoverflow.. I signed up via email and connected my Github account. Changed my email and my Github is still connected. I think we should do the same with bytedeck to retain the google account they used for logging in even if they changed their email. All emails they receive will go to the new one though.

Reply:
> "Choose “optional” or “none” to allow logins with an unverified e-mail address. " 

We want to allow logins with NO email address, but if there IS an address it needs to be verified.  Is that an option?

Unverified email addresses are a problem for lots of reasons. E.g You can put someone else's email addy, or you might have a typo and then we get rejected notification/announcement emails etc.

Reply:
> I think we were not talking about the same thing. My concern is existing users that have unverified emails in their User.email. There needs to be a way to migrate/convert/refactor these old accounts to the new system. We can't have two different email fields. All existing users in production will have this problem, so it needs to be resolved somehow.
> 
> Which of the two emails appears in their profile when they view it? When they change their email, which one is being changed?

When they view and change it, it's still the `User.email` that they will see and change.. I'm planning to create a post_save callback everytime a user changes their email to automatically send a confirmation email. In order to utilize allauth for sending confirmation emails, when the user changes their email, I just have to call `EmailAddress.objects.add_email(...).set_as_primary(...)`

Reply:
> We want to allow logins with NO email address, but if there IS an address it needs to be verified. Is that an option?
> 
> Unverified email addresses are a problem for lots of reasons. E.g You can put someone else's email addy, or you might have a typo and then we get rejected notification/announcement emails etc.

Unfortunately, this isn't an option but I'll try if I can hack this so we meet this requirement

Reply:
> > I think we were not talking about the same thing. My concern is existing users that have unverified emails in their User.email. There needs to be a way to migrate/convert/refactor these old accounts to the new system. We can't have two different email fields. All existing users in production will have this problem, so it needs to be resolved somehow.
> > 
> > Which of the two emails appears in their profile when they view it? When they change their email, which one is being changed?
> 
> When they view and change it, it's still the `User.email` that they will see and change.. I'm planning to create a post_save callback everytime a user changes their email to automatically send a confirmation email. In order to utilize allauth for sending confirmation emails, when the user changes their email, I just have to call `EmailAddress.objects.add_email(...).set_as_primary(...)`

So maybe to prevent confusion, anytime "email" is used in commits or docs or code comments or docstrings, it needs to specify which one: user.email or socialaccount.email? My preference would be to remove one from use completely.

What is the purpose of social account email? What is it currently used for? My understabding is that it's better than User.email because it comes with verification out of the box?

Reply:
It's only inside of this method that the `socialaccount.email` is used. This is the email we get from a social provider which is google in this case.

Then we just check if the one we got from google is being used as an email in `User.email`. If they do match, that's when we attempt to automatically connect a SocialAccount to a User.

Does that answer your questIon?